File: Java/src/_0946_Validate_Stack_Sequences/Solution.java
package _0946_Validate_Stack_Sequences;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Date: 2021/8/1 11:16
 * Content:
 * 执行用时：2 ms, 在所有 Java 提交中击败了92.64%的用户
 * 内存消耗：38.3 MB, 在所有 Java 提交中击败了14.38%的用户
 */
public class Solution {

    public boolean validateStackSequences(int[] pushed, int[] popped) {
        Deque<Integer> stk = new ArrayDeque<>();
        for (int i = 0, j = 0; i < pushed.length; i++) {
            stk.push(pushed[i]);
            while (!stk.isEmpty() && stk.peek() == popped[j]) {
                stk.pop();
                j++;
            }
        }
        return stk.isEmpty();
    }

    public static void main(String[] args) {

        int[]  pushed = {1, 2, 3, 4, 5}, poped = {4, 5, 3, 2, 1};
        System.out.println((new Solution()).validateStackSequences(pushed, poped));  // true
    }
}

File: Python/_0946_Validate_Stack_Sequences/Solution.py
# coding=utf-8
# Date: 2021/8/1 11:20
from typing import List


# 执行用时：44 ms, 在所有 Python3 提交中击败了61.98%的用户
# 内存消耗：15.1 MB, 在所有 Python3 提交中击败了37.23%的用户
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stk = []
        j = 0
        for i in range(len(pushed)):
            stk.append(pushed[i])
            while len(stk) != 0 and stk[len(stk) - 1] == popped[j]:
                stk.pop(len(stk) - 1)
                j += 1
        return len(stk) == 0


if __name__ == "__main__":
    pushed = [1, 2, 3, 4, 5]
    popped = [4, 5, 3, 2, 1]
    print(Solution().validateStackSequences(pushed, popped))  # True
